### PR TITLE
Reduced ops for grpc_chttp2_stream_map_find().

### PR DESCRIPTION
**Pull request by arjunroy**
Several asserts in grpc_chttp2_stream_map_find() can be converted to debug
asserts. This PR also templatizes the internal find() method to have it be
strict in the delete case (which saves some branches).

This is a subset of https://github.com/grpc/grpc/pull/19665 for easier review.

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/stream_map.cc [hunk 1]
```diff
@@ -74,7 +82,7 @@ void grpc_chttp2_stream_map_add(grpc_chttp2_stream_map* map, uint32_t key,
     } else {
       /* resize when less than 25% of the table is free, because compaction
          won't help much */
-      map->capacity = capacity = 3 * capacity / 2;
+      map->capacity = capacity = 2 * capacity;
```
Comment (yashykt) [inline src/core/ext/transport/chttp2/transport/stream_map.cc]: what's the reason for the change?

Comment (arjunroy) [reply src/core/ext/transport/chttp2/transport/stream_map.cc]: (Slightly) smaller number of ops when growing, and you'll need to grow less often since you grow a bit more.

File: src/core/ext/transport/chttp2/transport/stream_map.cc [hunk 2]
```diff
@@ -112,28 +122,28 @@ static void** find(grpc_chttp2_stream_map* map, uint32_t key) {
...
   }
 
+  GPR_DEBUG_ASSERT(!strict_find);
   return nullptr;
 }
 
 void* grpc_chttp2_stream_map_delete(grpc_chttp2_stream_map* map, uint32_t key) {
-  void** pvalue = find(map, key);
-  void* out = nullptr;
-  if (pvalue != nullptr) {
```
Comment (yashykt) [inline src/core/ext/transport/chttp2/transport/stream_map.cc]: is removing this if safe?

Comment (arjunroy) [reply src/core/ext/transport/chttp2/transport/stream_map.cc]: Yes, we should only be calling delete on things that we know we added otherwise it's an implementation bug that needs to be fixed. There is a GPR_DEBUG_ASSERT(pvalue != nullptr) I added to catch if it is null during development.

**Comment by yashykt** [comment]
Looking at the tests looks like something needs to be fixed

**Comment by arjunroy** [comment]
> Looking at the tests looks like something needs to be fixed

@yashykt That's the double deletion test that is failing, specifically because this change disallows double deletion. In my view that test should be removed and double deletion should not be supported (since this is not an externally visible generic data structure).
